### PR TITLE
FISH-8260 Document Evaluation of Extra Params in OpenID

### PR DESCRIPTION
**Pull request by aubi**
Documentation of change in 
https://github.com/payara/ecosystem-security-connectors/pull/348

### CODE REVIEW

File: docs/modules/ROOT/pages/Technical Documentation/Public API/OpenID Connect Support.adoc [hunk 1]
```diff
@@ -236,11 +236,18 @@ The annotation and MicroProfile properties has several configuration options. Th
 | N/A.
 
 | `extraParameters`
-|
+| `payara.security.openid.extraParams.raw` (compatibility with older version)
```
Comment (fturizo) [inline docs/modules/ROOT/pages/Technical Documentation/Public API/OpenID Connect Support.adoc:239]: What does the `compatibility with older version` sentence mean? Is this property new to `5.65.0` or is it the other way around?

File: docs/modules/ROOT/pages/Technical Documentation/Public API/OpenID Connect Support.adoc [hunk 2]
```diff
@@ -236,11 +236,18 @@ The annotation and MicroProfile properties has several configuration options. Th
...
 |
-| Must be in the form `key=value` and follow URL query pattern, i.e. `key1=value1&key2=value2&key2=value+with+spaces`. Key entries can be repeated
+| This is an array of strings in the form `key=value`, can use EL expression for values `key=#{expr}`. MP Config property is a string and must follow URL query pattern, i.e. `key1=value1&key2=value2&key2=value+with+spaces`. Key entries can be repeated.
+
+| `extraParametersExpression`
+| 
+| `false`
+| Overrides `extraParameters`, if specified, `extraParameters` is ignored. An array of extra options to be sent to the OpenID Connect provider.
+|
+| The content is supposed to be an EL expression. The expression must evaluate as either list of strings (apply the same rules as for `extraParameters`) or a string - this is split by comma and the strings are used as `extraParameters`. Example: `extraParametersExpression = "#{securityBean.extraParams}`
```
Comment (fturizo) [inline docs/modules/ROOT/pages/Technical Documentation/Public API/OpenID Connect Support.adoc:250]: Remove the ambiguity, instead of saying "The content is supposed to be an EL expression", state: "Corresponds to an EL expression string", for example.
Also, put the example on a new line to improve the cell's readability.

File: docs/modules/ROOT/pages/Technical Documentation/Public API/OpenID Connect Support.adoc [hunk 3]
```diff
@@ -268,6 +275,16 @@ then the MicroProfile Config property value always takes precedence over the `@O
 | false 
 |===
 
+=== `extraParameters` Evaluation
+
+The `extraParameters` option is syntactically an array of strings, so it cannot be directly injected by MP Config, which can only provide strings. There are ways, how to configure it:
```
Comment (fturizo) [inline docs/modules/ROOT/pages/Technical Documentation/Public API/OpenID Connect Support.adoc:280]: I'm a bit confused by this section and the change implemented, I have some questions:

. Isn't injecting a list/array of strings through MP Configuration technically possible? I recall that you can do this by using a custom converter. 
. Was the change implemented to the `payara.security.openid.extraParams` property? Did it support MP config injection before and now the `payara.security.openid.extraParams.raw` property supports it?

File: docs/modules/ROOT/pages/Technical Documentation/Public API/OpenID Connect Support.adoc [hunk 4]
```diff
@@ -268,6 +275,16 @@ then the MicroProfile Config property value always takes precedence over the `@O
...
+
+1. Backward compatibility: The annotation supports the MP Config value `payara.security.openid.extraParams.raw` in form of URL parameters. If this is specified, it takes precedence and evaluation stops.
+
+2. Expression-Based Configuration: The `extraParametersExpression` option was introduced for configuring arrays in a manner similar to Jakarta 10 and later. This option is evaluated as an EL expression. If it returns a list of strings, these strings are used in the form `"key=value"`. If the expression results in a single string, it must be in comma-separated format  (`"key1=value1,key2=value2"`). The values are split by commas and used. Any other type will result in an error.
```
Comment (fturizo) [inline docs/modules/ROOT/pages/Technical Documentation/Public API/OpenID Connect Support.adoc:284]: Remember that this is the Payara Plat 5.x documentation, so mentioning Jakarta EE 10 here doesn't make much sense. I would remove this sentence.

File: docs/modules/ROOT/pages/Technical Documentation/Public API/OpenID Connect Support.adoc [hunk 5]
```diff
@@ -268,6 +275,16 @@ then the MicroProfile Config property value always takes precedence over the `@O
...
+
+3. Default Configuration: If no other configuration is provided, `extraParameters` are resolved. Each string is split by `=`. The value can be either a direct value (`"key=value"`) or an expression (`"key=#{expression}"`).
+
```
Comment (fturizo) [inline docs/modules/ROOT/pages/Technical Documentation/Public API/OpenID Connect Support.adoc:287]: I'd add an example or 2 to clarify how these properties are expected to work in different scenarios since it is a bit confusing.

**Review by fturizo — CHANGES_REQUESTED** [review]
Needs some clarification on the changes made, see my comments. Also, @aubi, does this change only apply to Payara Plat 5, or does this PR need an equivalent one for 6.x?

File: docs/modules/ROOT/pages/Technical Documentation/Public API/OpenID Connect Support.adoc (hunk 1 above)
Comment (aubi) [reply docs/modules/ROOT/pages/Technical Documentation/Public API/OpenID Connect Support.adoc:239]: `payara.security.openid.extraParams.raw` is an older feature, for some reason not properly documented. The `extraParamsProperties` is a better solution, backported from similar annotation in Jakarta EE 10.

E.g. it is backported only to Payara 5, as Payara 6 (Jakarta EE 10) has the Jakarta annotation and there is no need for Payara-specific version.

File: docs/modules/ROOT/pages/Technical Documentation/Public API/OpenID Connect Support.adoc (hunk 4 above)
Comment (aubi) [reply docs/modules/ROOT/pages/Technical Documentation/Public API/OpenID Connect Support.adoc]: The request from the customer was "Do it similarly to JEE 10", so I thought it would be good to mention it here. I could use a work "backport".

File: docs/modules/ROOT/pages/Technical Documentation/Public API/OpenID Connect Support.adoc [hunk 6]
```diff
@@ -287,6 +294,49 @@ public class SecurityBean {
...
+There are several ways, how to configure `extraParameters` using Expression Language and Microprofile Config.
+
+==== Replaced by MP Config Value
+The annotation supports the MP Config value `payara.security.openid.extraParams.raw` (see `OpenIdAuthenticationDefinition.OPENID_MP_EXTRA_PARAMS_RAW`) in form of URL-encoded parameters. If specified, it takes precedence. Example of the value, e.g. in `META-INF/microprofile-config.properties` file:
+```
+payara.security.openid.extraParams.raw=key1=1%2C2&key2=value2
+```
+
+==== Replaced by `extraParametersExpression`
+The `extraParametersExpression` option was introduced in a manner similar to Jakarta 10 and later. If specified, this option is evaluated as an EL expression, e.g. it expects a bean returning the configuration. If it returns a list of strings, these strings are used in the form `"key=value"`. Values can use further expressions to be configurable at runtime.
```
Comment (Pandrex247) [inline docs/modules/ROOT/pages/Technical Documentation/Public API/OpenID Connect Support.adoc:309]: The product is Jakarta EE 10, not Jakarta 10
```suggestion
The `extraParametersExpression` option was introduced in a manner similar to Jakarta EE 10 and later. If specified, this option is evaluated as an EL expression, e.g. it expects a bean returning the configuration. If it returns a list of strings, these strings are used in the form `"key=value"`. Values can use further expressions to be configurable at runtime.
```

Comment (aubi) [reply docs/modules/ROOT/pages/Technical Documentation/Public API/OpenID Connect Support.adoc:309]: FISH-8260 jakarta "EE" forgotten

File: docs/modules/ROOT/pages/Technical Documentation/Public API/OpenID Connect Support.adoc [hunk 7]
```diff
@@ -236,11 +236,18 @@ The annotation and MicroProfile properties has several configuration options. Th
...
 | `false`
 | An array of extra options to be sent to the OpenID Connect provider.
 |
-| Must be in the form `key=value` and follow URL query pattern, i.e. `key1=value1&key2=value2&key2=value+with+spaces`. Key entries can be repeated
+| This is an array of strings in the form `key=value`, can use EL expression for values `key=#{expr}`. MP Config property is a string and must follow URL query pattern, i.e. `key1=value1&key2=value2&key2=value+with+spaces`. Key entries can be repeated. See the section <<extraparameters>> for more details.
+
+| `extraParametersExpression`
+| 
+| `false`
+| Overrides `extraParameters`, if specified, `extraParameters` is ignored. An array of extra options to be sent to the OpenID Connect provider.
```
Comment (Pandrex247) [inline docs/modules/ROOT/pages/Technical Documentation/Public API/OpenID Connect Support.adoc:248]: I'd reword this.
```suggestion
| An array of extra options to be sent to the OpenID Connect provider in the form of an Expression Language expression. If specified, overrides `extraParameters`. 
```

File: docs/modules/ROOT/pages/Technical Documentation/Public API/OpenID Connect Support.adoc [hunk 8]
```diff
@@ -287,6 +294,49 @@ public class SecurityBean {
 
 IMPORTANT: By default, the EL expressions are evaluated only once after the application is loaded and the evaluated values are then remembered until the application is reloaded, for performance reasons. This means that although the configuration can be evaluated dynamically the first time it's needed, it's not possible to change the configuration later on. If you need to dynamically modify the configuration during the lifetime of the application, follow the next section about multi-tenancy support.
 
+[[extraparameters]]
+=== Evaluation of `extraParameters`
+
+There are several ways, how to configure `extraParameters` using Expression Language and Microprofile Config.
+
+==== Replaced by MP Config Value
+The annotation supports the MP Config value `payara.security.openid.extraParams.raw` (see `OpenIdAuthenticationDefinition.OPENID_MP_EXTRA_PARAMS_RAW`) in form of URL-encoded parameters. If specified, it takes precedence. Example of the value, e.g. in `META-INF/microprofile-config.properties` file:
```
Comment (Pandrex247) [inline docs/modules/ROOT/pages/Technical Documentation/Public API/OpenID Connect Support.adoc:303]: > If specified, it takes precedence

Over what? The text in the configuration options table would indicate that `extraParametersExpression` takes precedence over this `extraParameters`.

I think this needs clarifying

File: docs/modules/ROOT/pages/Technical Documentation/Public API/OpenID Connect Support.adoc [hunk 9]
```diff
@@ -236,11 +236,18 @@ The annotation and MicroProfile properties has several configuration options. Th
...
 | `false`
 | An array of extra options to be sent to the OpenID Connect provider.
 |
-| Must be in the form `key=value` and follow URL query pattern, i.e. `key1=value1&key2=value2&key2=value+with+spaces`. Key entries can be repeated
+| This is an array of strings in the form `key=value`, can use EL expression for values `key=#{expr}`. MP Config property is a string and must follow URL query pattern, i.e. `key1=value1&key2=value2&key2=value+with+spaces`. Key entries can be repeated. See the section <<extraparameters>> for more details.
+
+| `extraParametersExpression`
+| 
+| `false`
+| An array of extra options to be sent to the OpenID Connect provider in the form of an Expression Language expression. If specified, overrides `extraParameters`.
```
Comment (Pandrex247) [inline docs/modules/ROOT/pages/Technical Documentation/Public API/OpenID Connect Support.adoc:248]:
```suggestion
| An array of extra options to be sent to the OpenID Connect provider in the form of an Expression Language expression. If specified, overrides `extraParameters` (but not `payara.security.openid.extraParams.raw`). 
```

File: docs/modules/ROOT/pages/Technical Documentation/Public API/OpenID Connect Support.adoc [hunk 10]
```diff
@@ -287,6 +294,49 @@ public class SecurityBean {
 
 IMPORTANT: By default, the EL expressions are evaluated only once after the application is loaded and the evaluated values are then remembered until the application is reloaded, for performance reasons. This means that although the configuration can be evaluated dynamically the first time it's needed, it's not possible to change the configuration later on. If you need to dynamically modify the configuration during the lifetime of the application, follow the next section about multi-tenancy support.
 
+[[extraparameters]]
+=== Evaluation of `extraParameters`
+
+There are several ways, how to configure `extraParameters` using Expression Language and Microprofile Config.
```
Comment (Pandrex247) [inline docs/modules/ROOT/pages/Technical Documentation/Public API/OpenID Connect Support.adoc:300]:
```suggestion
There are several ways, how to configure `extraParameters` using Expression Language and MicroProfile Config. The order of precedence is: `payara.security.openid.extraParams.raw` overrides `extraParametersExpression`, which overrides the plain `extraParams` value.
```

**Review by Pandrex247 — APPROVED** [review]

**Review by fturizo — APPROVED** [review]
Improved the wording and structured based on your examples, @aubi.
